src/main/java/net/minecraftforge/common/JigsawCategory.java

```diff
@@ -0,0 +1,145 @@
+package net.minecraftforge.common;
+
+import com.google.common.collect.Lists;
+import com.google.common.collect.Maps;
+import com.mojang.datafixers.util.Pair;
+import net.minecraft.util.ResourceLocation;
+import net.minecraft.world.gen.feature.jigsaw.JigsawPiece;
+import java.util.List;
+import java.util.Map;
+import java.util.Random;
+
+public class JigsawCategory {
+    private final List<Pair<JigsawPiece, Integer>> pieces;
+    private final int categoryWeight;
+    private final ResourceLocation registryName;
+    private int totalPieceWeight;
+
+    public JigsawCategory(ResourceLocation registryName, int categoryWeight, List<Pair<JigsawPiece, Integer>> pieces) {
+        this.pieces = Lists.newArrayList(pieces);
+        this.categoryWeight = categoryWeight;
+        this.registryName = registryName;
+        this.totalPieceWeight = pieces.stream().mapToInt(Pair::getSecond).sum();
+    }
+
+    public List<Pair<JigsawPiece, Integer>> getPieces()
+    {
+        return pieces;
+    }
+
+    public ResourceLocation getRegistryName()
+    {
+        return registryName;
+    }
+
+    public int getCategoryWeight()
+    {
+        return categoryWeight;
+    }
+
+    public int getTotalPieceWeight()
+    {
+        return totalPieceWeight;
+    }
+
+    public void recalculateWeight()
+    {
+        this.totalPieceWeight = pieces.stream().mapToInt(Pair::getSecond).sum();
+    }
+
+    public static class PieceItem
+    {
+        private final float weight;
+        private final JigsawPiece piece;
+
+        public PieceItem(float itemWeightIn, JigsawPiece piece)
+        {
+            this.weight = itemWeightIn;
+            this.piece = piece;
+        }
+
+        public JigsawPiece getPiece()
+        {
+            return piece;
+        }
+
+        public float getWeight()
+        {
+            return weight;
+        }
+    }
+
+    public static class WeightedFloatRandom
```

Comment:
Please add a comment to day this is a version of the vanilla WeigthedRandom with floats